infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -333,9 +334,17 @@ private void applyShadowStream(UUID streamId, long snapshot) {
         int numBatches = 1;
 
         for (SMREntry smrEntry : smrEntries) {
+            // Apply all SMR entries in a single transaction as long as it does not exceed the max write size(25MB).
+            // It was observed that special streams(ProtobufDescriptor table), can get a lot of updates, especially
+            // due to schema updates during an upgrade.  If the table was not checkpointed and trimmed on the Source,
+            // no de-duplication on these updates will occur.  As a result, the transaction size can be large.
+            // Although it is within the maxWriteSize limit, deserializing these entries to read the table can cause an
+            // OOM on applications running with a small memory footprint.  So for such tables, introduce an
+            // additional limit of max number of entries(50 by default) applied in a single transaction.  This
+            // algorithm is in line with the limits imposed in Compaction and Restore workflows.
             if (bufferSize + smrEntry.getSerializedSize() >
-                logReplicationMetadataManager.getRuntime().getParameters()
-                    .getMaxWriteSize()) {
+                logReplicationMetadataManager.getRuntime().getParameters().getMaxWriteSize() ||
```

Comment:
too many operator in the same line? Also not sure about the operator precedence order. would you like to add parenthesis to increase the readability?

Reply:
the parts before and after the 'or' operator are evaluated separately.  In this case, it is self-explanatory IMO and codacy complains if the parentheses are unnecessary.  We can discuss offline if you disagree.